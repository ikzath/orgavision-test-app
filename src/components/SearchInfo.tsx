import React, { useState, useEffect } from 'react';
import { Container, Row, Col, Button } from 'reactstrap';
import DisplayCard from './DisplayCard';
import axios, { AxiosResponse } from "axios";
import IPost from "../interfaces";
import "../style/SearchInfo.css";
import DisplayLoadingCard from "./DisplayLoadingCard";


const SearchInfo: React.FC = () => {
    
    const defaultProps:IPost[] = [];
    const [query, setquery] = useState<string>();
    const [posts, setPosts]: [IPost[], (posts: IPost[]) => void] = React.useState(defaultProps);
    const [loading, setLoading]: [boolean, (loading: boolean) => void] = React.useState<boolean>(false);
    const [searchResults, setsearchResults]: [IPost[], (searchResults: IPost[]) => void] = React.useState(defaultProps);
    const BarStyling = {width:"20rem",background:"#F2F1F9", border:"none", padding:"0.5rem", borderRadius: "15px", textDecoration: "none"};


    const handleChange = (
      e: React.ChangeEvent<HTMLInputElement>
    ) => {
      setquery(e.currentTarget.value)
      if(query && query.length > 1) {
        if(query.length % 2 === 0){
          // setTimeout(() => fetchData(), 2000)
          fetchData();
        }            
      }
    };

      // get dynamic results with search nput 
      useEffect(() => {  
        fetchData();

      }, [query])


    useEffect(() => {
        const fetchAllData = () => {
        setLoading(true)
        axios.get(
              'https://orgavision-codingchallenge.azurewebsites.net/v2/article',
          )
          .then(( res: AxiosResponse<any>)=> {
            setLoading(false)
            setPosts(res.data.records);
            console.log(posts);
          })   
          .catch((err: any) => {
            console.log('error', err);
          })
        };    
          fetchAllData();    
    }, []);
  
  
  const fetchData = async(): Promise<any> => {
    try{
        setLoading(true)
        const result = await axios.get(`https://orgavision-codingchallenge.azurewebsites.net/v2/article?search=${query}`
        );        
        setLoading(false)
        let results = result.data;
        setsearchResults(results.records);
        console.log(results);
    }
    catch (err: any) {                        
      console.log('no results', err);
    }
    };


    let queryResults =         
      searchResults.map((item: any, i: number)=> (
        <Col md={3} key={item.id}> 
            <DisplayCard
            key={item.id}
            id = {item.id}
            title={item.fields.title}
            teaser={item.fields.teaser}
            category= {item.fields.category.map((itemCategory: any, i: number) => (
            <Button style={{ color: 'whitesmoke', backgroundColor: '#A0A0A0'}} to='' key={i} onClick={()=> setquery(itemCategory)}>{itemCategory}</Button> 
                ))} />
        </Col>
        ))     


      let pageOnLoadResults =                 
        posts.map((item: any, i: number)=> (
        <Col md={3} key={item.id} > 
            <DisplayCard 
                key={item.id}
                id = {item.id}
                title={item.fields.title}
                teaser={item.fields.teaser}
                category= {item.fields.category.map((itemCategory: any, i: number) => (
                <Button style={{ color: 'whitesmoke', backgroundColor: '#A0A0A0'}} to='' key={i} onClick={()=> setquery(itemCategory)}>{itemCategory}</Button> 
                ))} />
        </Col>
    ))                  

            
return (
  <div className='searchInfoDisplay'>
  <div className='nav'>
      <div className="nav__text">
          <h3>Knowledge Base</h3>
      </div>     
      <div className="nav__text2">
          <h3>KB</h3>
      </div>   
      <div className="nav__search">
      <input 
          style={BarStyling}
          type="text"
          placeholder="Suchen..."
          value={query || ''}
          onChange={handleChange}
          />
          </div>       
      </div>
      <div>
      <Container>
          {loading ? <DisplayLoadingCard /> :
      <Row>
          {query ? queryResults : pageOnLoadResults }
      </Row> }
      </Container>        
      </div>
      </div>
    )
}

export default SearchInfo
