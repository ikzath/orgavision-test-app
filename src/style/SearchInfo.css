/* .searchInfoDisplay {
  background-color: rgb(221, 213, 213);
} */

.nav {
  position: sticky;
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex: 1;
  color: gray;
  background-color: #303094;
  height: 9vh;
  top: 0;
  z-index: 100;
}

.nav__text {
  color: white;
  margin-left: 10px;
}

.nav__text2 {
  color: white;
  margin-left: 15px;
  display: none;
}

.nav__search {
  margin-right: 10px;
  border: none;
}

/* responsive sizing */
@media only screen and (min-device-width: 320px) and (max-device-width: 480px) and (-webkit-min-device-pixel-ratio: 2) {
  .nav__text {
    display: none;
  }
  .nav__text2 {
    display: block;
  }

  /* .nav__search > input {
      width: 5rem;
    } */
}
