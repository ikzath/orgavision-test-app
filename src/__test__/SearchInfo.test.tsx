import SearchInfo from "../components/SearchInfo";
import React from "react";
import * as ReactDOM from "react-dom";
import { fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";


describe('SearchInfo component tests', ()=> {
    let container: HTMLDivElement;

    beforeEach(()=> {
        container = document.createElement('div');
        document.body.appendChild(container)
        ReactDOM.render(<SearchInfo />, container)
    })

    afterEach(()=> {
        document.body.removeChild(container)
        container.remove();
    })

    it('renders initial document correctly', ()=> {
        const inputs = container.querySelectorAll('input');
        expect(inputs).toHaveLength(1);
        expect(inputs[0].placeholder).toBe('Suchen...')
    })

    it('it passes credentials correctly',()=> {
        const inputs = container.querySelectorAll('input');
        const onChangrInput = inputs[0];
        fireEvent.change(onChangrInput, { target: { value:'something'}});
    })
})

    it('it matches snapshot', ()=> {
        const tree = renderer.create(<SearchInfo/>).toJSON();
        expect(tree).toMatchSnapshot();
    })



