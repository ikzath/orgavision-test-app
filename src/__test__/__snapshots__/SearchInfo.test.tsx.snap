// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<div
  className="searchInfoDisplay"
>
  <div
    className="nav"
  >
    <div
      className="nav__text"
    >
      <h3>
        Knowledge Base
      </h3>
    </div>
    <div
      className="nav__text2"
    >
      <h3>
        KB
      </h3>
    </div>
    <div
      className="nav__search"
    >
      <input
        onChange={[Function]}
        placeholder="Suchen..."
        style={
          Object {
            "background": "#F2F1F9",
            "border": "none",
            "borderRadius": "15px",
            "padding": "0.5rem",
            "textDecoration": "none",
            "width": "20rem",
          }
        }
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    data-testid="loading"
  >
    <div
      className="container"
    >
      <div
        className="row"
      />
    </div>
  </div>
</div>
`;
